pkgs/tools/security/pinentry/default.nix

```diff
@@ -1,60 +1,96 @@
 { fetchurl, fetchpatch, stdenv, lib, pkgconfig, autoreconfHook
 , libgpgerror, libassuan
-, libcap ? null, libsecret ? null, ncurses ? null, gtk2 ? null, gcr ? null
-, qt4 ? null, qt5 ? null
-, enableEmacs ? false
+, ncurses, gtk2, qt
+, libcap ? null, libsecret ? null, gcr ? null
+, enabledFlavors ? [ "curses" "tty" "gtk2" "qt" "gnome3" "emacs" ]
 }:
 
-assert qt5 != null -> qt4 == null;
-assert qt4 != null -> qt5 == null;
+with stdenv.lib;
+
+assert isList enabledFlavors && enabledFlavors != [];
 
 let
   mkDerivation =
-    if qt5 != null
-      then qt5.mkDerivation
+    if (builtins.elem "qt" enabledFlavors)
+      then qt.mkDerivation
       else stdenv.mkDerivation;
+
+  mkFlag = pfxTrue: pfxFalse: cond: name:
+    "--${if cond then pfxTrue else pfxFalse}-${name}";
+  mkEnable = mkFlag "enable" "disable";
+  mkWith = mkFlag "with" "without";
+
+  mkEnablePinentry = f:
+    let
+      info = flavorInfo.${f};
+      inputs = info.buildInputs or [];
+      flag = flavorInfo.${f}.flag or null;
+      inputsSatisfied = all (f: ! isNull f) inputs;
+    in
+      assert inputsSatisfied;
+      optionalString (flag != null)
+        (mkEnable (elem f enabledFlavors && inputsSatisfied) ("pinentry-" + flag));
+
+  flavorInfo = {
+    curses = { bin = "curses"; buildInputs = [ ncurses ]; };
+    tty = { bin = "tty"; flag = "tty"; };
```

Comment:
I think it might make sense to just put pinentry-tty and pinentry-emacs in the default output. Perhaps pinentry-curses as well. They are all pretty small and don't increase closures sizes.

Reply:
Currently, the first enabled flavor will land in `$out/bin/pinentry`.

We could introduce a `alwaysOut` boolean attribute in `flavorInfo`, which will add a symlink from `$flavour/bin/pinentry-$flavor` to `$flavor/bin/pinentry-$flavor`, but I'm not sure if it's worth the effort - does this guarantee the right pinentry is picked?

Reply:
After thinking about it a bit, I'd like to keep the current behaviour, and not introduce any defaults or "always outputs". If somebody wishes to change the pinentry to be used, they can set `enabledFlavors`, or just pick one of the outputs already provided in top-level for convenience.